_delb/nodes.py

```diff
@@ -3006,17 +3078,342 @@ def not_wrapper(node: NodeBase) -> bool:
+                        # hence there's a default namespace in use
+                        self.__new_namespace_declaration(namespace)
+                    else:
+                        self._prefixes[None] = ""
+                    continue
+
+                try:
+                    self._namespaces._fallback = node.namespaces
+                    prefix = self._namespaces.lookup_prefix(namespace)
+
+                    if prefix is None:
+                        if "" in self._prefixes.values():
+                            # hence there's an empty namespace in use
+                            self._prefixes[namespace] = ""
+                            namespace = None
+                            raise KeyError
+                        else:
+                            prefix = ""
+                    else:
+                        prefix = prefix + ":"
+
+                    self._prefixes[namespace] = prefix
+
+                except KeyError:
+                    assert isinstance(namespace, str)
+                    self.__new_namespace_declaration(namespace)
+
+    def _init_config(
+        self,
+        align_attributes: bool,
+        indentation: str,
+        namespaces: None | NamespaceDeclarations,
+        text_width: int,
+    ):
+        self.align_attributes = align_attributes
+        self.indentation = indentation
+
+        if namespaces is None:
+            namespaces = {}
+        self._namespaces = Namespaces(namespaces)
+
+        self.text_width = text_width
+        self.__text_wrapper = TextWrapper(
+            width=text_width,
+            expand_tabs=False,
+            replace_whitespace=True,
+            drop_whitespace=True,
+            initial_indent="",
+            subsequent_indent="",
+            fix_sentence_endings=False,
+            break_long_words=False,
+            break_on_hyphens=False,
+        )
+
+    def __new_namespace_declaration(self, namespace: None | str):
+        for i in range(2**16):
+            prefix = f"ns{i}:"
+            if prefix not in self._prefixes.values():
+                self._prefixes[namespace] = prefix
+                return
+        else:
+            raise NotImplementedError("Just don't.")
+
+    def __generate_attributes_data(self, node: TagNode) -> dict[str, str]:
+        data = {}
+
+        for attribute in (node.attributes[a] for a in sorted(node.attributes)):
+            assert isinstance(attribute, Attribute)
+            data[
+                self._prefixes[attribute.namespace] + attribute.local_name
+            ] = f'''"{self.sanitize_text(attribute.value).replace('"', "&quot;")}"'''
+
+        return data
+
+    @staticmethod
+    def sanitize_text(text: str) -> str:
+        return text.replace("&", "&amp;").replace(">", "&gt;").replace("<", "&lt;")
+
+    def __serialize_aligned_attributes(self, data: dict[str, str]):
+        key_width = max(len(k) for k in data)
+        for key, value in data.items():
```

Comment:
it just occurred to me that this would be even funnier:

```suggestion
        key_width, value_width = (max(len(item[p]) for item in d.items()) for p in range(2))
        for key, value in sorted(
            data.items(), key=lambda item: len(item[0]) + len(item[1]) / value_width
        ):
```

Reply:
(would require that we acount for non-string values and empty values obviously)

Reply:
could you please wrap that up in a commit including extended tests. i'd like to play with that. i like the idea, but am considering the footprint as well. a separate sorting function is preferable over a lambda expression. shall the parameter name be changed?

as a matter of fact: most characters are valid (parameter) names in Python!

Reply:
and i mean, with a lot of attributes there could be a Hain.

Reply:
ok

Reply:
yeah, I just realized... why is that?

Reply:
yes, obviously!